[Regression] "Improve vm support for ref types" commit generates invalid C code

Hello,

After updating to the latest Nim devel yesterday, I started seeing this C compilation error for the `nim-emacs-module` project:

```
CC: stdlib_ospaths
CC: stdlib_osproc
Error: execution of an external compiler program 'gcc -c  -w -I/home/kmodi/usr_local/apps/6/emacs/master/bin -std=gnu99 -fPIC  -I/home/kmodi/stow/pkgs/nim/bad/lib -o /home/kmodi/stow/pub_dotfiles/emacs/dot-emacs.d/misc/nim-emacs-module/test/nimcache/emacs_module_sample.o /home/kmodi/stow/pub_dotfiles/emacs/dot-emacs.d/misc/nim-emacs-module/test/nimcache/emacs_module_sample.c' failed with exit code: 1

/home/kmodi/stow/pub_dotfiles/emacs/dot-emacs.d/misc/nim-emacs-module/test/nimcache/emacs_module_sample.c: In function ‘emacs_module_init’:
/home/kmodi/stow/pub_dotfiles/emacs/dot-emacs.d/misc/nim-emacs-module/test/nimcache/emacs_module_sample.c:924:0: error: unterminated argument list invoking macro "DEFUN"


/home/kmodi/stow/pub_dotfiles/emacs/dot-emacs.d/misc/nim-emacs-module/test/nimcache/emacs_module_sample.c:130:1: error: ‘DEFUN’ undeclared (first use in this function)
 DEFUN ("DEFUN ("sample-mod-test-return-t", nimEmacs_sample_mod_test_return_t, 1, 1, NULL, NULL);
 ^~~~~
/home/kmodi/stow/pub_dotfiles/emacs/dot-emacs.d/misc/nim-emacs-module/test/nimcache/emacs_module_sample.c:130:1: note: each undeclared identifier is reported only once for each function it appears in
/home/kmodi/stow/pub_dotfiles/emacs/dot-emacs.d/misc/nim-emacs-module/test/nimcache/emacs_module_sample.c:130:1: error: expected ‘;’ at end of input
/home/kmodi/stow/pub_dotfiles/emacs/dot-emacs.d/misc/nim-emacs-module/test/nimcache/emacs_module_sample.c:130:1: error: expected declaration or statement at end of input
make: *** [Makefile:24: sample.so] Error 1
```

By git bisect, I narrowed down to this commit to cause this regression: https://github.com/nim-lang/Nim/commit/236bc06b5f1a1a1fc4a711007aed0b60b6656a4f.

I also verified that reverting just this commit fixes the regression.

## Recipe

**You do not need to have Emacs installed to reproduce this issue.**

```
git clone https://github.com/kaushalmodi/nim-emacs-module
cd nim-emacs-module
git checkout origin/recipe-for-ref-type-vm-support-commit-caused-regression   
cd test
make sample
```

**You need to be on a Nim devel build built using https://github.com/nim-lang/Nim/commit/236bc06b5f1a1a1fc4a711007aed0b60b6656a4f or later.**

- [`recipe-for-ref-type-vm-support-commit-caused-regression` branch][4]
- [Generated C code **with** the 236bc06 commit -- bad][1]
- [Generated C code **after reverting** the 236bc06 commit -- good][2]
- [Diff of the two][3]

/cc @GULPF 

[1]: https://gist.github.com/kaushalmodi/caea6a290aae3f7209adfd3a1288d053#file-emacs_module_sample__bug-c
[2]: https://gist.github.com/kaushalmodi/caea6a290aae3f7209adfd3a1288d053#file-emacs_module_sample__with_236bc06b_reverted__working-c
[3]: https://gist.github.com/kaushalmodi/caea6a290aae3f7209adfd3a1288d053#file-working_vs_buggy-diff
[4]: https://github.com/kaushalmodi/nim-emacs-module/tree/recipe-for-ref-type-vm-support-commit-caused-regression

--- a/compiler/vm.nim
+++ b/compiler/vm.nim
@@ -209,7 +209,7 @@ proc writeField(n: var PNode, x: TFullReg) =
   of rkNone: discard
   of rkInt: n.intVal = x.intVal
   of rkFloat: n.floatVal = x.floatVal
-  of rkNode: n = x.node
+  of rkNode: n = copyValue(x.node)
   of rkRegisterAddr: writeField(n, x.regAddr[])
   of rkNodeAddr: n = x.nodeAddr[]
 
@@ -912,6 +912,7 @@ proc rawExecute(c: PCtx, start: int, tos: PStackFrame): TFullReg =
       if a.kind == nkSym:
         regs[ra].node = if a.sym.ast.isNil: newNode(nkNilLit)
                         else: copyTree(a.sym.ast)
+        regs[ra].node.flags.incl nfIsRef
       else:
         stackTrace(c, tos, pc, "node is not a symbol")
     of opcEcho:
@@ -1462,6 +1463,7 @@ proc rawExecute(c: PCtx, start: int, tos: PStackFrame): TFullReg =
       else:
         regs[ra].node = newNodeI(nkIdent, c.debug[pc])
         regs[ra].node.ident = getIdent(c.cache, regs[rb].node.strVal)
+        regs[ra].node.flags.incl nfIsRef
     of opcSetType:
       if regs[ra].kind != rkNode:
         internalError(c.config, c.debug[pc], "cannot set type")

--- a/tests/vm/tvmmisc.nim
+++ b/tests/vm/tvmmisc.nim
@@ -90,3 +90,33 @@ block:
       var some = newStringOfCap(size)
       result = size
     doAssert f(4) == 4
+
+# #7871
+static:
+  type Obj = object
+    field: int
+  var s = newSeq[Obj](1)
+  var o = Obj()
+  s[0] = o
+  o.field = 2
+  doAssert s[0].field == 0
+
+# #8125
+static:
+   let def_iter_var = ident("it")
+
+# #8142
+static:
+  type Obj = object
+    names: string
+
+  proc pushName(o: var Obj) =
+    var s = ""
+    s.add("FOOBAR")
+    o.names.add(s)
+
+  var o = Obj()
+  o.names = ""
+  o.pushName()
+  o.pushName()
+  doAssert o.names == "FOOBARFOOBAR"

